src/hyp3_isce2/burst.py

```diff
@@ -0,0 +1,238 @@
+    return {'url': url}
+
+
+def wait_for_extractor(response: requests.Response, sleep_time: int = 15) -> bool:
+    if response.status_code == 202:
+        time.sleep(15)
+        return False
+
+    response.raise_for_status()
+    return True
+
+
+def download_from_extractor(asf_session: requests.Session, burst_params: BurstParams, content: str):
+    burst_request = create_burst_request(burst_params, content=content)
+    burst_request['cookies'] = {'asf-urs': asf_session.cookies['asf-urs']}
+
+    for ii in range(1, 11):
+        print(f'Download attempt #{ii}')
+        response = asf_session.get(**burst_request)
+        downloaded = wait_for_extractor(response)
+        if downloaded:
+            break
+
+    if not downloaded:
+        raise RuntimeError('Download failed too many times')
+
+    return response.content
+
+
+def download_metadata(asf_session: requests.Session, burst_params: BurstParams, out_file: Union[Path, str] = None):
+    content = download_from_extractor(asf_session, burst_params, 'metadata')
+    metadata = ET.fromstring(content)
+
+    if not out_file:
+        return metadata
+
+    with open(out_file, 'wb') as f:
+        f.write(content)
+
+    return str(out_file)
+
+
+def download_burst(asf_session: requests.Session, burst_params: BurstParams, out_file: Union[Path, str]):
+    content = download_from_extractor(asf_session, burst_params, 'geotiff')
+
+    with open(out_file, 'wb') as f:
+        f.write(content)
+
+    return str(out_file)
+
+
+def spoof_safe(asf_session: requests.Session, burst: BurstMetadata, base_path: Path = Path('.')) -> Path:
+    """Creates this file structure:
+    SLC.SAFE/
+    ├── manifest.safe
+    ├── measurement/
+    │   └── burst.tif
+    └── annotation/
+        ├── annotation.xml
+        └── calbiration/
+            ├── calibration.xml
+            └── noise.xml
+    """
+    safe_path = base_path / f'{burst.safe_name}.SAFE'
+    annotation_path = safe_path / 'annotation'
+    calibration_path = safe_path / 'annotation' / 'calibration'
+    measurement_path = safe_path / 'measurement'
+    paths = [annotation_path, calibration_path, measurement_path]
+    for p in paths:
+        if not p.exists():
+            p.mkdir(parents=True)
+
+    et_args = {'encoding': 'UTF-8', 'xml_declaration': True}
+
+    ET.ElementTree(burst.annotation).write(annotation_path / burst.annotation_name, **et_args)
+    ET.ElementTree(burst.calibration).write(calibration_path / burst.calibration_name, **et_args)
+    ET.ElementTree(burst.noise).write(calibration_path / burst.noise_name, **et_args)
+    ET.ElementTree(burst.manifest).write(safe_path / 'manifest.safe', **et_args)
+
+    burst_params = BurstParams(burst.safe_name, burst.swath, burst.polarization, burst.burst_number)
+    download_burst(asf_session, burst_params, measurement_path / burst.measurement_name)
+
+    return safe_path
+
+
+# TODO currently only validated for descending orbits
```

Comment:
Is this blocking?

Reply:
I've added a new test for this but it's failing 😢. I'll work on fixing!

Reply:
Fixed!